### DISCUSSION TITLE
Thoughts on "Tailor-made Concurrency" wiki section

### DISCUSSION THREAD

**Discussion by scottyeager** [python-telegram-bot/python-telegram-bot#4438] [Q&A]
I'm current doing my homework before migrating one of my bots from v13 to v21. Part of that was reading the concurrency wiki [page](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Concurrency), which has a [section](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Concurrency#tailor-made-concurrency) on "Tailor-made Concurrency".

While the example does a good job of showing how async code can suffer from race conditions, the way that the solution shown is described was a bit confusing to me.

Since I have limited async experience, I thought this discussion format might be a good way to bring clarity. So, here are my thoughts and questions 🙂

The gist of this example is that we can make the code safe by moving awaited coroutines into tasks. Since we eliminated all `await`s from the critical path, there's no chance that another coroutine can run and change the shared state between our read and write steps.

These tasks are generated from a single async function with two steps. First is a synchronous logging step (no await). Second is an async message send that's awaited.

My first question then is, why put the logging function into the task? Since it's not awaited it can't cause a race condition. On the other hand, since it's not awaitable, it can't run concurrently in a native way with async. You'd need to run it in an executor.

Now my second question is, when are these tasks actually executed?

The suggestion is that the tasks will run concurrently with the ongoing `transaction` function. Especially as it's written here:

> the log entries will be in order most of the time, assuming the database operations take long enough for the log to complete.

But the thing is that these tasks can't run until there's an opportunity to do so presented by either an `await` (which we've eliminated) or the end of the `transaction` coroutine.

It looks to me like all of the logging and sending of messages will happen only after the transaction processing is completed. We only have a single thread of execution and creating tasks in itself does not pass control back to the scheduler.

That is, unless we're talking about [eager](https://docs.python.org/3/library/asyncio-task.html#id9) task execution. In that case, the logging function would run synchronously and then the task would get scheduled at the point of blocking on sending the message. But I didn't find any mention of eager execution in the article nor any evidence that it's used by default in PTB.

Am I missing something?

**Comment by Bibo-Joshi** [comment]
Hi. Let me try to give my personal take on your questions:

> My first question then is, why put the logging function into the task? Since it's not awaited it can't cause a race condition. On the other hand, since it's not awaitable, it can't run concurrently in a native way with async. You'd need to run it in an executor.

Mainly consistency, I'd say. If you keep the logging in `transaction` directly, but move the call to `bot.send_message` to a task, then the timestamps of the log & the message can differ more significantly. But this is not the ultimate truth and you could ofc also argue that the log should be strictly at the places where the actual transactions happen, hence keep it in `transaction`. You also right that `bank.log` is not awaitable so that having it in a task doesn't give any IO speedup - but it also doesn't slow the program down :)

> Now my second question is, when are these tasks actually executed?

> It looks to me like all of the logging and sending of messages will happen only after the transaction processing is completed. We only have a single thread of execution and creating tasks in itself does not pass control back to the scheduler.

You're probably right with the second point and this can then also be a reason on why one could decide to keep `bank.log` within `transaction` directly. The situation changes if `bank.write_account` and `bank.read_account` need to be awaited as well (see the other note below the example).
In the end, the important thing about `asyncio` IMO is that you don't have to worry about when tasks are actually executed. If you care about order for something, don't put it in a task. Otherwise, it's "fire and forget" and the event loop will take care of it :)

After all, it's just a simple exapmle and ofc can't address all real life concerns …

Hope that my reply is helpful for you in some way :)
